osquery/sql/sqlite_string.cpp

```diff
@@ -210,6 +210,90 @@ static void regexStringMatchFunc(sqlite3_context* context,
                       SQLITE_TRANSIENT);
 }
 
+/**
+ * @brief Concatenate strings ignoring nulls
+ */
+static void concatStringFunc(sqlite3_context* context,
+                             int argc,
+                             sqlite3_value** argv) {
+  // Nothing to concat, no point.
+  if (argc == 0) {
+    return;
+  }
+
+  std::string output;
+
+  for (auto i = 0; i < argc; i++) {
+    if (SQLITE_NULL == sqlite3_value_type(argv[i])) {
+      continue;
+    }
+
+    output.append(reinterpret_cast<const char*>(sqlite3_value_text(argv[i])));
+  }
+
+  // Give up if the output is so large it's length overflows int
+  if (output.size() > std::numeric_limits<int>::max()) {
+    LOG(INFO) << "Too much data for concat";
+    sqlite3_result_error(context, "Too much data for concat", -1);
+    return;
+  }
+
+  sqlite3_result_text(context,
+                      output.c_str(),
+                      static_cast<int>(output.size()),
+                      SQLITE_TRANSIENT);
+}
+
+/**
+ * @brief Concatenate strings ignoring nulls, with separator
+ */
+static void concatWSStringFunc(sqlite3_context* context,
+                               int argc,
+                               sqlite3_value** argv) {
+  // First arg is the separator. Error if not present
+  if (argc == 0) {
+    LOG(INFO) << "Missing separator to concat_ws";
+    sqlite3_result_error(context, "Missing separator to concat_ws", -1);
+    return;
+  }
+
+  const std::string sep =
+      (SQLITE_NULL != sqlite3_value_type(argv[0]))
+          ? reinterpret_cast<const char*>(sqlite3_value_text(argv[0]))
+          : "";
+
+  // If we have less than 2 args, no point to iterating. Return an empty string
+  if (argc == 1) {
+    return;
+  }
+
+  std::string output;
+
+  for (auto i = 1; i < argc; i++) {
+    if (SQLITE_NULL == sqlite3_value_type(argv[i])) {
+      continue;
+    }
+
+    output.append(reinterpret_cast<const char*>(sqlite3_value_text(argv[i])));
+
+    if (i + 1 < argc) {
+      output.append(sep);
+    }
+  }
+
+  // Give up if the output is so large it's length overflows int
+  if (output.size() > std::numeric_limits<int>::max()) {
+    LOG(INFO) << "Too much data for concat_ws";
+    sqlite3_result_error(context, "Too much data for concat_ws", -1);
+    return;
+  }
```

Comment:
This is probably best to be `SQLITE_MAX_LENGTH`.

Reply:
Hrm, that does seem more correct, but I don't see how. AFAICT `SQLITE_MAX_LENGTH` isn't a constant, but instead it's an argument to `sqlite3_limit` which needs to be called with the `db`. Which isn't available in this function. If you've got an example I'm happy to change it.

Reply:
The docs indicate it's set as a preprocessory macro. Does replacing `std::numeric_limits<int>::max()` with `SQLITE_MAX_LENGTH` work?

Reply:
```
/Users/seph/checkouts/osquery/osquery/osquery/sql/sqlite_string.cpp:238:23: error: use of undeclared identifier 'SQLITE_MAX_LENGTH'
  if (output.size() > SQLITE_MAX_LENGTH {
                      ^
1 error generated.
```

There is also `SQLITE_MAX_LENGTH` which is a macro, and compiles. But is also set to 0. 

All the docs I can find are about using `sqlite3_limit ` -- https://www.sqlite.org/c3ref/limit.html